### PR TITLE
[Model] Scene Graph Extraction Model with GluonCV

### PR DESCRIPTION
**Pull request by hetong007**
## Description
This PR reproduces [Graphical Contrastive Losses for Scene Graph Parsing](https://arxiv.org/pdf/1903.02728.pdf) with GluonCV. In this example we provide 

- Data preparation scripts
- Training and Validation pipeline
- Demo Jupyter notebook
- Link to original paper and implementation

It is large and self-contained, and probably the first example of computer vision in DGL.
Not sure who are the best reviewers, so @jermainewang could you help?

P.S. This PR also provides the original example to reproduce https://github.com/dmlc/dgl/issues/1130

## Checklist
Please feel free to remove inapplicable items for your PR.
- [X] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [X] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [X] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

**Comment by jermainewang** [comment]
As the PR is quite large. Several suggestions before we could move on to the review.
1. Clean up the files. Remove non-code files such as png, jpg and ipy notebook.
2. Create a readme with how to run, dependencies and the expected results. Please refer to the other examples.
3. It looks like you made some visualization (which is cool!), consider put the result in the readme too. You could consider creating a GIF.
4. Since there are many files, please add some explanations in the readme about the purpose of each file as a guideline for people to understand your codes. Elaborate more details in the module docstring in each file, and even more specifications in the docstring of core functions.

**Comment by hetong007** [comment]
Hi @jermainewang , thank you for the input. 

1. PNG/JPG files serve the purpose of ipynb input and result visualization. Is `dmlc/web-data` a better place to host? IMHO ipy notebook works as a great demo, as people may also need a demo to showcase the prediction other than numbers from metrics. Do we have guidelines of ipynb files?
2. Please check [here](https://github.com/hetong007/dgl/tree/scenegraphinit/examples/mxnet/scenegraph) for the how to run, dependencies and the expected results.
3. It is not a video application so cannot guarantee any results with GIF. Probably I can include a few more images. Besides, it comes with the same issue of the first item: where do we host the images?
4. Good suggestion. I'll add more details in readme and docstrings.

**Comment by hetong007** [comment]
Update: 

1. The ipynb turned into a demo script, images are moved to `dmlc/web-data`.
2. README improved.
3. Docstrings added

File: examples/mxnet/scenegraph/demo_reldn.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+import argparse
+import mxnet as mx
+import gluoncv as gcv
+from gluoncv.data.transforms import presets
+from model import faster_rcnn_resnet101_v1d_custom, RelDN
+from utils import *
+from data import *
+
+def parse_args():
+    parser = argparse.ArgumentParser(description='Train Faster-RCNN networks e2e.')
```
Comment (jermainewang) [inline examples/mxnet/scenegraph/demo_reldn.py]: Is this a script just for detection or scene graph extraction? Also the Faster-RCNN net is not really trained right? Look like it is pre-trained instead.

File: examples/mxnet/scenegraph/train_faster_rcnn.sh [hunk 2]
```diff
@@ -0,0 +1,2 @@
+MXNET_CUDNN_AUTOTUNE_DEFAULT=0 CUDNN_AUTOTUNE_DEFAULT=0 MXNET_GPU_MEM_POOL_TYPE=Round MXNET_GPU_MEM_POOL_ROUND_LINEAR_CUTOFF=28 python train_faster_rcnn.py \
+    --gpus 0,1,2,3,4,5,6,7 --dataset visualgenome -j 60 --batch-size 8 --val-interval 20 --save-prefix faster_rcnn_resnet101_v1d_visualgenome/
```
Comment (jermainewang) [inline examples/mxnet/scenegraph/train_faster_rcnn.sh]: Could you confirm that the script works with one gpu as well?

File: examples/mxnet/scenegraph/train_freq_prior.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
+import numpy as np
+import json, pickle, os, argparse
+
+def parse_args():
+    parser = argparse.ArgumentParser(description='Train Faster-RCNN networks e2e.')
```
Comment (jermainewang) [inline examples/mxnet/scenegraph/train_freq_prior.py]: Description is not correct.

File: examples/mxnet/scenegraph/train_reldn.py [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+from mxnet import nd, gluon
+from gluoncv.data.batchify import Pad
+from gluoncv.utils import makedirs
+
+from model import faster_rcnn_resnet101_v1d_custom, RelDN
+from utils import *
+from data import *
+
+def parse_args():
+    parser = argparse.ArgumentParser(description='Train Faster-RCNN networks e2e.')
```
Comment (jermainewang) [inline examples/mxnet/scenegraph/train_reldn.py]: ditto

File: examples/mxnet/scenegraph/validate_reldn.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+import logging, time, argparse
+from mxnet import nd, gluon
+from gluoncv.data.batchify import Pad
+
+from model import faster_rcnn_resnet101_v1d_custom, RelDN
+from utils import *
+from data import *
+
+def parse_args():
+    parser = argparse.ArgumentParser(description='Train Faster-RCNN networks e2e.')
```
Comment (jermainewang) [inline examples/mxnet/scenegraph/validate_reldn.py]: ditto

File: examples/mxnet/scenegraph/README.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+After the training, one can evaluate the results with multiple commonly-used metrics:
+
+```
+bash validate_reldn.sh
+```
+
+## Demo
+
+We provide a demo notebook of running the model with real-world pictures. Be aware that you need trained model to generate meaningful results from the demo.
```
Comment (jermainewang) [inline examples/mxnet/scenegraph/README.md]: How to run the demo?

Comment (jermainewang) [reply examples/mxnet/scenegraph/README.md]: Maybe just put one pic here showing the result? You could upload the image to imgur.

File: examples/mxnet/scenegraph/data/prepare_visualgenome.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    # move all images into folder `VG_100K`
+    vg_100k_path = os.path.join(path, 'VG_100K')
+    vg_100k_2_path = os.path.join(path, 'VG_100K_2')
+    files_2 = os.listdir(vg_100k_2_path)
+    for fl in files_2:
+        shutil.move(os.path.join(vg_100k_2_path, fl),
+                    os.path.join(vg_100k_path, fl))
+
+def download_json(path, overwrite=False):
+    url = 'https://drive.google.com/uc?id=1VDuba95vIPVhg5DiriPtwuVA6mleYGad'
```
[the remark this answers is not in the document]
Comment (hetong007) [reply examples/mxnet/scenegraph/data/prepare_visualgenome.py]: I'd love to. Do we have a public S3 bucket or sth equivalent for that?

File: examples/mxnet/scenegraph/demo_reldn.py (hunk 1 above)
Comment (hetong007) [reply examples/mxnet/scenegraph/demo_reldn.py]: The Faster-CNN net needs to be trained. I could put the pretrained models to the same S3 bucket along with the dataset.

**Comment by hetong007** [comment]
My current implementation is faster with one GPU. However, at this moment, I cannot train on more than 1 GPU with MXNet so the absolute execution time is longer.

**Comment by jermainewang** [comment]
> I cannot train on more than 1 GPU with MXNet so the absolute execution time is longer.

Even using horovod?

**Comment by hetong007** [comment]
> > I cannot train on more than 1 GPU with MXNet so the absolute execution time is longer.
> 
> Even using horovod?

I haven't tried horovod yet, may worth trying after this PR.

File: examples/mxnet/scenegraph/data/prepare_visualgenome.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jermainewang) [reply examples/mxnet/scenegraph/data/prepare_visualgenome.py]: Call @VoVAllen for help.

**Comment by hetong007** [comment]
@jermainewang  Comments addressed. Would you please take another look and, if everything looks good, help to merge it? Thanks!

**Comment by hetong007** [comment]
@jermainewang May I have your approval, or feedback on other issues? Thanks.

**Comment by jermainewang** [comment]
> @jermainewang May I have your approval, or feedback on other issues? Thanks.

I've already approved. Go ahead and merge it. Thanks!

**Pull request merged**